cob_cartesian_controller/src/trajectory_interpolator/trajectory_interpolator.cpp

```diff
@@ -27,136 +27,140 @@
+        path = linear_path;
+    }
 
     // Interpolate the linear path
-    for(int i = 0 ; i < path_matrix[0].size() ; i++)
+    for(unsigned int i = 0 ; i < path_matrix[0].size() ; i++)
     {
-        if(move_lin.rotate_only)
+        if(linear_path.back() == 0)
         {
-            pose.position.x = move_lin.start.position.x;
-            pose.position.y = move_lin.start.position.y;
-            pose.position.z = move_lin.start.position.z;
+            pose.position.x = as.move_lin.start.position.x;
+            pose.position.y = as.move_lin.start.position.y;
+            pose.position.z = as.move_lin.start.position.z;
         }
         else
         {
-            pose.position.x = move_lin.start.position.x + linear_path.at(i) * (move_lin.end.position.x - move_lin.start.position.x) / Se;
-            pose.position.y = move_lin.start.position.y + linear_path.at(i) * (move_lin.end.position.y - move_lin.start.position.y) / Se;
-            pose.position.z = move_lin.start.position.z + linear_path.at(i) * (move_lin.end.position.z - move_lin.start.position.z) / Se;
+            pose.position.x = as.move_lin.start.position.x + linear_path.at(i) * (as.move_lin.end.position.x - as.move_lin.start.position.x) / linear_path.back();
+            pose.position.y = as.move_lin.start.position.y + linear_path.at(i) * (as.move_lin.end.position.y - as.move_lin.start.position.y) / linear_path.back();
+            pose.position.z = as.move_lin.start.position.z + linear_path.at(i) * (as.move_lin.end.position.z - as.move_lin.start.position.z) / linear_path.back();
         }
 
-        // Transform RPY to Quaternion
-        q.setRPY(roll_path.at(i), pitch_path.at(i), yaw_path.at(i));
-        tf::quaternionTFToMsg(q, pose.orientation);
-
+        tf::quaternionTFToMsg(q_start.slerp(q_end, path.at(i) * norm_factor), pose.orientation);
         pose_array.poses.push_back(pose);
     }
     return true;
 }
 
 bool TrajectoryInterpolator::circularInterpolation(geometry_msgs::PoseArray& pose_array,
-                                                   cob_cartesian_controller::MoveCircStruct& move_circ)
+                                                   const cob_cartesian_controller::CartesianActionStruct& as)
 {
     pose_array.header.stamp = ros::Time::now();
-    pose_array.header.frame_id = root_frame_;
-
-    tf::Quaternion q;
-    tf::Transform C, P, T;
-
-    std::vector<double> path_array;
-    geometry_msgs::Pose pose;
-
-    double Se = move_circ.end_angle - move_circ.start_angle;
-
-    bool forward;
-    if(Se < 0)
-    {
-        forward = false;
-    }
-    else
-    {
-        forward = true;
-    }
-    Se = std::fabs(Se);
+     pose_array.header.frame_id = root_frame_;
+
+     tf::Quaternion q;
+     tf::Transform C, P, T;
+
+     std::vector<double> path_array;
+     std::vector<double> path_matrix[4];
+
+     geometry_msgs::Pose pose;
+
+     double Se = as.move_circ.end_angle - as.move_circ.start_angle;
+
+     bool forward;
```

Comment:
What does "forward" mean? clockwise vs. counter-clockwise?
Does it have something to do with shortest angle?